linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+                    else
+                    {
+                        result->task_array_count = THREAD_LIMIT;
+                        result->list_index = 0;
+                        if (sem_init(&result->semaphore, 0 , 0) != 0)
+                        {
+                            LogError("Failure creating semi_init");
+                        }
+                        else
+                        {
+                            (void)interlocked_exchange(&result->state, THREADPOOL_STATE_NOT_OPEN);
+                            result->thread_count = 0;
+                            result->task_count = 0;
+                            (void)interlocked_exchange(&result->quitting, 0);
+
+                            (void)interlocked_exchange_64(&result->insert_idx, 0);
+                            (void)interlocked_exchange_64(&result->consume_idx, 0);
+
+                            goto all_ok;
+                        }
+                        srw_lock_destroy(result->srw_lock);
+                    }
+                    free(result->task_array);
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            if (ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care) != THREADAPI_OK)
+            {
+                LogError("Failure joining thread");
+            }
+        }
+
+        free(threadpool->task_array);
+        free(threadpool->thread_handle_list);
+
+        sem_destroy(&threadpool->semaphore);
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (interlocked_add(&threadpool->state, 0) != (int32_t)THREADPOOL_STATE_OPEN)
+        {
+            LogWarning("Threadpool schedule work called on non-open pool");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            //int64_t insert_pos = (interlocked_increment_64(&threadpool->insert_idx) % threadpool->task_array_count) - 1;
+            int64_t insert_pos = interlocked_exchange_64(&threadpool->insert_idx, threadpool->insert_idx+1) % threadpool->task_array_count ;
```

Comment:
>nt64_t insert_pos = interlocked_exchange_64(&threadpool->insert_idx, threadpool->insert_idx+1) % threadpool->task_array_count ; [](http://example.com/codeflow?start=13&length=127)

missing a loop, we can't read an interlocked like this and need to handle multiple schedule calls at the same time.